File: 1. Creational Design Patterns/Factory Method/Go/main.go
package main

import (
    "fmt"
    "os"
)

// Product

type Transport interface {
    Deliver()
}

type Truck struct{}

func (Truck) Deliver() {
    fmt.Println("Deliver by land in a box.")
}

type Ship struct{}

func (Ship) Deliver() {
    fmt.Println("Deliver by sea in a container.")
}

// Creator

type Logistics interface {
    PlanDelivery()
}

type baseLogistics struct{}

func (baseLogistics) PlanDelivery() {}

// Concrete Creators

type RoadLogistics struct{}

func (RoadLogistics) PlanDelivery() { Truck{}.Deliver() }

type SeaLogistics struct{}

func (SeaLogistics) PlanDelivery() { Ship{}.Deliver() }

func main() {
    // Optional arg: "sea" to use Ship; any other value (or none) uses Truck
    var l Logistics
    if len(os.Args) > 1 && os.Args[1] == "sea" {
        l = SeaLogistics{}
    } else {
        l = RoadLogistics{}
    }
    l.PlanDelivery()
}

File: 1. Creational Design Patterns/Factory Method/Rust/main.rs
// Factory Method – Rust Example
trait Transport {
    fn deliver(&self);
}

struct Truck;
impl Transport for Truck {
    fn deliver(&self) {
        println!("Deliver by land in a box.");
    }
}

struct Ship;
impl Transport for Ship {
    fn deliver(&self) {
        println!("Deliver by sea in a container.");
    }
}

trait Logistics {
    fn create_transport(&self) -> Box<dyn Transport>;
    fn plan_delivery(&self) {
        self.create_transport().deliver();
    }
}

struct RoadLogistics;
impl Logistics for RoadLogistics {
    fn create_transport(&self) -> Box<dyn Transport> { Box::new(Truck) }
}

struct SeaLogistics;
impl Logistics for SeaLogistics {
    fn create_transport(&self) -> Box<dyn Transport> { Box::new(Ship) }
}

fn main() {
    let log: Box<dyn Logistics> = if std::env::args().any(|a| a == "sea") {
        Box::new(SeaLogistics)
    } else {
        Box::new(RoadLogistics)
    };

    log.plan_delivery();
}
